src/openms/include/OpenMS/FILTERING/ID/IDFilter.h

```diff
@@ -1002,6 +1106,204 @@ namespace OpenMS
+    static void keepBestPerPeptide(std::vector<PeptideIdentification>& pep_ids, bool ignore_mods, bool ignore_charges, Size nr_best_spectrum)
+    {
+      annotateBestPerPeptide(pep_ids, ignore_mods, ignore_charges, nr_best_spectrum);
+      HasMetaValue<PeptideHit> best_per_peptide{"best_per_peptide", "true"};
+      keepMatchingItemsUnroll(pep_ids, best_per_peptide);
+    }
+
+    static void keepBestPerPeptidePerRun(std::vector<ProteinIdentification>& prot_ids, std::vector<PeptideIdentification>& pep_ids, bool ignore_mods, bool ignore_charges, Size nr_best_spectrum)
+    {
+      annotateBestPerPeptidePerRun(prot_ids, pep_ids, ignore_mods, ignore_charges, nr_best_spectrum);
+      HasMetaValue<PeptideHit> best_per_peptide{"best_per_peptide", "true"};
+      keepMatchingItemsUnroll(pep_ids, best_per_peptide);
+    }
+
+    template <class MapType>
+    static void keepBestPerPeptidePerRun(MapType& prot_and_pep_ids, bool ignore_mods, bool ignore_charges, Size nr_best_spectrum)
+    {
+      const auto& prot_ids = prot_and_pep_ids.getProteinIdentifications();
+
+      std::map<std::string, std::unordered_map<std::string, std::map<Int, PeptideHit*>>> best_peps_per_run;
+      for (const auto& idrun : prot_ids)
+      {
+        best_peps_per_run[idrun.getIdentifier()] = std::unordered_map<std::string, std::map<Int, PeptideHit*>>();
+      }
+
+      for (auto& feat : prot_and_pep_ids)
+      {
+        annotateBestPerPeptidePerRunWithData(best_peps_per_run, feat.getPeptideIdentifications(), ignore_mods, ignore_charges, nr_best_spectrum);
+      }
+
+      annotateBestPerPeptidePerRunWithData(best_peps_per_run, prot_and_pep_ids.getUnassignedPeptideIdentifications(), ignore_mods, ignore_charges, nr_best_spectrum);
+
+      HasMetaValue<PeptideHit> best_per_peptide{"best_per_peptide", "true"};
+      keepMatchingPeptideHits(prot_and_pep_ids, best_per_peptide);
+    }
+
+    static void annotateBestPerPeptidePerRun(const std::vector<ProteinIdentification>& prot_ids, std::vector<PeptideIdentification>& pep_ids, bool ignore_mods, bool ignore_charges, Size nr_best_spectrum)
+    {
+      std::map<std::string, std::unordered_map<std::string, std::map<Int, PeptideHit*>>> best_peps_per_run;
+      for (const auto& id : prot_ids)
+      {
+        best_peps_per_run[id.getIdentifier()] = std::unordered_map<std::string, std::map<Int, PeptideHit*>>();
+      }
+      annotateBestPerPeptidePerRunWithData(best_peps_per_run, pep_ids, ignore_mods, ignore_charges, nr_best_spectrum);
+    }
+
+    /// Annotates PeptideHits from PeptideIdentification if it is the best peptide hit for its peptide sequence
+    /// Adds metavalue "bestForItsPeps" which can be used for additional filtering.
+    static void annotateBestPerPeptidePerRunWithData(std::map<std::string, std::unordered_map<std::string, std::map<Int, PeptideHit*>>>& best_peps_per_run, std::vector<PeptideIdentification>& pep_ids, bool ignore_mods, bool ignore_charges, Size nr_best_spectrum)
```

Comment:
this would be so much easier to read with a typedef - please use them! or even a small class! how could I know which map relate to the run, which one to the peptide etc? like this its easy to get confused ...